internal/controller/datadogagent/feature/otelcollector/feature.go

```diff
@@ -0,0 +1,169 @@
+package otelcollector
+
+import (
+	"strconv"
+	"strings"
+
+	apicommon "github.com/DataDog/datadog-operator/api/datadoghq/common"
+	"github.com/DataDog/datadog-operator/api/datadoghq/v2alpha1"
+	apiutils "github.com/DataDog/datadog-operator/api/utils"
+	"github.com/DataDog/datadog-operator/internal/controller/datadogagent/feature"
+	"github.com/DataDog/datadog-operator/internal/controller/datadogagent/feature/otelcollector/defaultconfig"
+	"github.com/DataDog/datadog-operator/internal/controller/datadogagent/object/configmap"
+	"github.com/DataDog/datadog-operator/internal/controller/datadogagent/object/volume"
+	"github.com/DataDog/datadog-operator/pkg/kubernetes"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	otelAgentVolumeName = "otel-agent-config-volume"
+	otelConfigFileName  = "otel-config.yaml"
+)
+
+func init() {
+	err := feature.Register(feature.OtelAgentIDType, buildOtelCollectorFeature)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func buildOtelCollectorFeature(options *feature.Options) feature.Feature {
+	return &otelCollectorFeature{}
+}
+
+type otelCollectorFeature struct {
+	customConfig  *v2alpha1.CustomConfig
+	owner         metav1.Object
+	configMapName string
+	ports         []*corev1.ContainerPort
+}
+
+func (o otelCollectorFeature) ID() feature.IDType {
+	return feature.OtelAgentIDType
+}
+
+func (o *otelCollectorFeature) Configure(dda *v2alpha1.DatadogAgent) feature.RequiredComponents {
+	o.owner = dda
+	if dda.Spec.Features.OtelCollector.Conf != nil {
+		o.customConfig = dda.Spec.Features.OtelCollector.Conf
+	}
+	o.configMapName = v2alpha1.GetConfName(dda, o.customConfig, v2alpha1.DefaultOTelAgentConf)
+
+	if len(dda.Spec.Features.OtelCollector.Ports) == 0 {
+		dda.Spec.Features.OtelCollector.Ports = []*corev1.ContainerPort{
+			{
+				Name:          "otel-http",
+				ContainerPort: 4318,
+				HostPort:      4318,
+				Protocol:      corev1.ProtocolTCP,
+			},
+			{
+				Name:          "otel-grpc",
+				ContainerPort: 4317,
+				HostPort:      4317,
+				Protocol:      corev1.ProtocolTCP,
+			},
+		}
+	}
+	o.ports = dda.Spec.Features.OtelCollector.Ports
```

Comment:
Why not assign the configured ports' structure directly to `o.ports`? As a general practice, we don't modify DDA withing the features, instead save relevant config in feature struct `otelCollectorFeature`.

Reply:
Makes sense, updated in: https://github.com/DataDog/datadog-operator/pull/1559/commits/ff7868ebd53219d2ebb593197edf50c5f2584bbf.